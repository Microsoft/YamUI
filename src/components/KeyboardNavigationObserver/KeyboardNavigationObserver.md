### Examples

Basic usage:

```js { "props": { "data-description": "basic" } }
const { GutterSize, TextSize } = require('../Block');

const renderInKeyboardMode = () => {
  return (
    <div style={{ backgroundColor: '#adedba', transition: 'background-color 1s' }}>
      <Block padding={GutterSize.XXLARGE} textSize={TextSize.LARGE}>Currently navigating via keyboard</Block>
    </div>
  );
};

const renderInMouseMode = (inView) => {
  return (
    <div style={{ backgroundColor: '#f98985', transition: 'background-color 1s' }}>
      <Block padding={GutterSize.XXLARGE} textSize={TextSize.LARGE}>Currently navigating via mouse</Block>
    </div>
  );
};

<KeyboardNavigationObserver
  renderInKeyboardMode={renderInKeyboardMode}
  renderInMouseMode={renderInMouseMode}
  onEnterKeyboardMode={action('entered keyboard mode')}
  onEnterMouseMode={action('entered mouse mode')}
/>
```
