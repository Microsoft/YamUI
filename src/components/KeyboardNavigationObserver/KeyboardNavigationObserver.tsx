/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import {
  getMode,
  subscribe,
  unsubscribe,
  NavigationMode,
  NavigationModeCallback,
} from '../../util/accessibility/keyboardNavigation';

export { NavigationMode, NavigationModeCallback };

export interface KeyboardNavigationObserverProps extends BaseComponentProps {
  /**
   * A callback which will be triggered when the user begins navigating via keyboard.
   */
  onEnterKeyboardMode?: () => void;

  /**
   * A callback which will be triggered when the user begins navigating via mouse.
   */
  onEnterMouseMode?: () => void;

  /**
   * Render prop to return child content when user is navigating via keyboard.
   */
  renderInKeyboardMode?: () => React.ReactElement<any>;

  /**
   * Render prop to return child content when user is navigating via mouse.
   */
  renderInMouseMode?: () => React.ReactElement<any>;
}

export interface KeyboardNavigationObserverState {
  mode: NavigationMode;
}

/**
 * KeyboardNavigationObserver allows for conditional child rendering and callbacks depending on whether the user
 * is currently navigating via mouse or keyboard. This is useful for features like infinite scroll, where
 * new pages should be loaded by button-click instead of scroll-based auto-loaded, ensuring keyboard navigation
 * users are able to navigate past the end of the content section and into footer or sidebar sections.
 */
export default class KeyboardNavigationObserver extends React.Component<
  KeyboardNavigationObserverProps,
  KeyboardNavigationObserverState
> {
  constructor(props: KeyboardNavigationObserverProps) {
    super(props);
    this.state = {
      mode: getMode(),
    };
  }

  public render() {
    const { renderInKeyboardMode, renderInMouseMode } = this.props;
    const { mode } = this.state;

    if (mode === 'keyboard' && renderInKeyboardMode) {
      return renderInKeyboardMode();
    }
    if (mode === 'mouse' && renderInMouseMode) {
      return renderInMouseMode();
    }

    return null;
  }

  public componentDidMount() {
    subscribe(this.onModeChange);
  }

  public componentWillUnmount() {
    unsubscribe(this.onModeChange);
  }

  private onModeChange = (mode: NavigationMode) => {
    const { onEnterKeyboardMode, onEnterMouseMode } = this.props;

    if (mode === 'keyboard' && onEnterKeyboardMode) {
      onEnterKeyboardMode();
    }
    if (mode === 'mouse' && onEnterMouseMode) {
      onEnterMouseMode();
    }

    this.setState({ mode });
  };
}
