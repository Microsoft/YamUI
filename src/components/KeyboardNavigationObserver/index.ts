/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export { default } from './KeyboardNavigationObserver';
export * from './KeyboardNavigationObserver.types';
