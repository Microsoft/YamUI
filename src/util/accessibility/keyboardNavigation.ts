/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import KeyboardNavigationEventListener, {
  NavigationMode,
  NavigationModeCallback,
} from './KeyboardNavigationEventListener';

export { NavigationMode, NavigationModeCallback };

let navListener: KeyboardNavigationEventListener;
document.addEventListener('DOMContentLoaded', () => {
  navListener = new KeyboardNavigationEventListener(document.body);
});

export const getMode = (): NavigationMode => {
  return navListener.getMode();
};

export const subscribe = (callback: NavigationModeCallback): void => {
  navListener.subscribe(callback);
};

export const unsubscribe = (callback: NavigationModeCallback): void => {
  navListener.unsubscribe(callback);
};
