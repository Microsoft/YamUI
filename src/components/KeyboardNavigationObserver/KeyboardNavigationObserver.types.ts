/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { BaseComponentProps } from '../../util/BaseComponent/props';
import { NavigationMode, NavigationModeCallback } from '../../util/accessibility/keyboardNavigation';
export { NavigationMode, NavigationModeCallback };

export interface KeyboardNavigationObserverProps extends BaseComponentProps {
  /**
   * A callback which will be triggered when the user begins navigating via keyboard.
   */
  onEnterKeyboardMode?: (() => void);

  /**
   * A callback which will be triggered when the user begins navigating via mouse.
   */
  onEnterMouseMode?: (() => void);

  /**
   * Render prop to return child content when user is navigating via keyboard.
   */
  renderInKeyboardMode?: (() => React.ReactNode);

  /**
   * Render prop to return child content when user is navigating via mouse.
   */
  renderInMouseMode?: (() => React.ReactNode);
}

export interface KeyboardNavigationObserverState {
  mode: NavigationMode;
}
