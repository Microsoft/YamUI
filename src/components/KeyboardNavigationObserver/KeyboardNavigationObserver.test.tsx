/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
// jest.mock('../../util/accessibility/keyboardNavigation', () => ({
//   getMode: () => 'mouse',
//   subscribe: (callback: any) => callback('mouse'),
//   unsubscribe: () => 'noop',
// }));

import * as React from 'react';
import { mount, shallow, ReactWrapper, ShallowWrapper } from 'enzyme';
import KeyboardNavigationObserver, {
  KeyboardNavigationObserverProps,
  KeyboardNavigationObserverState,
} from './index';
import * as keyboardNavigation from '../../util/accessibility/keyboardNavigation';

describe('<KeyboardNavigationObserver />', () => {
  let component: ShallowWrapper<KeyboardNavigationObserverProps, KeyboardNavigationObserverState>;
  let mounted: ReactWrapper<KeyboardNavigationObserverProps, KeyboardNavigationObserverState>;

  const renderInKeyboardMode = () => <span>KEYBOARD</span>;
  const renderInMouseMode = () => <span>MOUSE</span>;

  let onEnterKeyboardModeCallback: jest.Mock<{}>;
  let onEnterMouseModeCallback: jest.Mock<{}>;

  beforeEach(() => {
    onEnterKeyboardModeCallback = jest.fn();
    onEnterMouseModeCallback = jest.fn();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('in mouse mode', () => {
    beforeEach(() => {
      jest.spyOn(keyboardNavigation, 'getMode').mockReturnValue('mouse');
    });

    describe('with minimal options', () => {
      beforeEach(() => {
        component = shallow(<KeyboardNavigationObserver />);
      });

      it('matches its snapshot', () => {
        expect(component).toMatchSnapshot();
      });
    });

    describe('with all options', () => {
      beforeEach(() => {
        component = shallow(
          <KeyboardNavigationObserver
            renderInKeyboardMode={renderInKeyboardMode}
            renderInMouseMode={renderInMouseMode}
            onEnterKeyboardMode={onEnterKeyboardModeCallback}
            onEnterMouseMode={onEnterMouseModeCallback}
          />,
        );
      });

      it('matches its snapshot', () => {
        expect(component).toMatchSnapshot();
      });
    });

    describe('when mounted and mode is switched to keyboard', () => {
      beforeEach(() => {
        jest
          .spyOn(keyboardNavigation, 'subscribe')
          .mockImplementation((callback: any) => callback('keyboard'));

        mounted = mount(
          <KeyboardNavigationObserver
            renderInKeyboardMode={renderInKeyboardMode}
            renderInMouseMode={renderInMouseMode}
            onEnterKeyboardMode={onEnterKeyboardModeCallback}
            onEnterMouseMode={onEnterMouseModeCallback}
          />,
        );
      });

      afterEach(() => {
        if (mounted) {
          mounted.unmount();
        }
      });

      it('triggers its onEnterKeyboardMode callback', () => {
        expect(onEnterKeyboardModeCallback).toHaveBeenCalled();
      });

      it('matches its snapshot', () => {
        expect(mounted).toMatchSnapshot();
      });
    });
  });

  describe('in keyboard mode', () => {
    beforeEach(() => {
      jest.spyOn(keyboardNavigation, 'getMode').mockReturnValue('keyboard');
    });

    describe('with all options', () => {
      beforeEach(() => {
        component = shallow(
          <KeyboardNavigationObserver
            renderInKeyboardMode={renderInKeyboardMode}
            renderInMouseMode={renderInMouseMode}
            onEnterKeyboardMode={onEnterKeyboardModeCallback}
            onEnterMouseMode={onEnterMouseModeCallback}
          />,
        );
      });

      it('matches its snapshot', () => {
        expect(component).toMatchSnapshot();
      });
    });

    describe('when mounted and mode is switched to mouse', () => {
      beforeEach(() => {
        jest
          .spyOn(keyboardNavigation, 'subscribe')
          .mockImplementation((callback: any) => callback('mouse'));

        mounted = mount(
          <KeyboardNavigationObserver
            renderInKeyboardMode={renderInKeyboardMode}
            renderInMouseMode={renderInMouseMode}
            onEnterKeyboardMode={onEnterKeyboardModeCallback}
            onEnterMouseMode={onEnterMouseModeCallback}
          />,
        );
      });

      afterEach(() => {
        if (mounted) {
          mounted.unmount();
        }
      });

      it('triggers its onEnterMouseMode callback', () => {
        expect(onEnterMouseModeCallback).toHaveBeenCalled();
      });

      it('matches its snapshot', () => {
        expect(mounted).toMatchSnapshot();
      });
    });
  });
});
