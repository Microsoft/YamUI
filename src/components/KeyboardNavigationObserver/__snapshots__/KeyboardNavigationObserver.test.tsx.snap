// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KeyboardNavigationObserver /> in keyboard mode when mounted and mode is switched to mouse matches its snapshot 1`] = `
<KeyboardNavigationObserver
  onEnterKeyboardMode={[MockFunction]}
  onEnterMouseMode={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
    }
  }
  renderInKeyboardMode={[Function]}
  renderInMouseMode={[Function]}
>
  <span>
    MOUSE
  </span>
</KeyboardNavigationObserver>
`;

exports[`<KeyboardNavigationObserver /> in keyboard mode with all options matches its snapshot 1`] = `
<span>
  KEYBOARD
</span>
`;

exports[`<KeyboardNavigationObserver /> in mouse mode when mounted and mode is switched to keyboard matches its snapshot 1`] = `
<KeyboardNavigationObserver
  onEnterKeyboardMode={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
    }
  }
  onEnterMouseMode={[MockFunction]}
  renderInKeyboardMode={[Function]}
  renderInMouseMode={[Function]}
>
  <span>
    KEYBOARD
  </span>
</KeyboardNavigationObserver>
`;

exports[`<KeyboardNavigationObserver /> in mouse mode with all options matches its snapshot 1`] = `
<span>
  MOUSE
</span>
`;

exports[`<KeyboardNavigationObserver /> in mouse mode with minimal options matches its snapshot 1`] = `null`;
